server/jetstream_benchmark_test.go

```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+			nc, _ = jsClientConnectURL(b, clientUrl)
+			defer nc.Close()
+		}
+
+		// random bytes as payload
+		messageData := make([]byte, messageSize)
+		if _, err := rand.Read(messageData); err != nil {
+			b.Fatal(err)
+		}
+
+		// waits for all publishers sub-routines and for main thread to be ready
+		var publishersReadyWg sync.WaitGroup
+		publishersReadyWg.Add(numPubs + 1)
+
+		// create N publishers
+		publishers := func() []BenchPublisher {
+			pubs := make([]BenchPublisher, numPubs)
+			for i := range pubs {
+				// create a quit channel
+				pubs[i].quitCh = make(chan bool, 1)
+				// create publisher connection and jetstream context
+				ncPub, err := nats.Connect(clientUrl)
+				if err != nil {
+					b.Fatal(err)
+				}
+				pubs[i].conn = ncPub
+				jsPub, err := ncPub.JetStream()
+				if err != nil {
+					b.Fatal(err)
+				}
+				pubs[i].js = jsPub
+
+				// publisher successfully initialized
+				publishersReadyWg.Done()
+			}
+			return pubs
+		}()
+
+		// wait group to ensure all publishers have been torn down
+		var finishedPublishersWg sync.WaitGroup
+		finishedPublishersWg.Add(numPubs)
+
+		// wait group blocks main thread until publish workload is completed, it is decremented after stream receives b.N messages from all publishers
+		var benchCompleteWg sync.WaitGroup
+		benchCompleteWg.Add(1)
+
+		// start go routines for all publishers, wait till all publishers are initialized before starting publish workload
+		for i := range publishers {
+			go func(pubId int) {
+				// signal that this publisher has been torn down
+				defer finishedPublishersWg.Done()
+
+				// wait till all other publishers are ready to start workload
+				publishersReadyWg.Wait()
+
+				// publish until quitCh is closed
+				for {
+					select {
+					case <-publishers[pubId].quitCh:
+						publishers[pubId].conn.Close()
+						return
+					default:
+						// continue publishing
+					}
+					pubAck, err := publishers[pubId].js.Publish(subject, messageData)
+					if err != nil {
+						publishers[pubId].publishErrors += 1
+						fmt.Println(err)
+						continue
+					}
+
+					// all messages have been published to stream
+					if int(pubAck.Sequence) == b.N {
+						benchCompleteWg.Done()
+					}
+				}
+			}(i)
+		}
+
+		// set bytes per operation
+		b.SetBytes(messageSize)
+		// main thread is ready
+		publishersReadyWg.Done()
+		// wait till publishers are ready
+		publishersReadyWg.Wait()
+
+		// start the clock
+		b.ResetTimer()
+		// wait till termination cond reached
+		benchCompleteWg.Wait()
+		// stop the clock
+		b.StopTimer()
+
+		// send quit signal to all publishers
+		for i := range publishers {
+			publishers[i].quitCh <- true
+		}
+		// wait for all publishers to shutdown
+		finishedPublishersWg.Wait()
+
+		// sum errors from all publishers
+		totalErrors := 0
+		for _, publisher := range publishers {
+			totalErrors += publisher.publishErrors
+		}
+		errorRate := 100 * float64(totalErrors) / float64(b.N)
```

Comment:
The denominator here is not exactly right. Should be number of total calls to `Publish`

```
N 3
Publish calls 10
Errors 7

Current error rate: 230%
Actual error rate: 70%
```